xop.go

```diff
@@ -79,86 +79,152 @@ func (d *xopDecoder) getXopContentIDIncludePath(element *etree.Element, path []s
+		// - the name of the field type
+		fieldName := ""
+		if fieldName = getNameFromTag(tag); fieldName == "" {
+			if fieldName = getExplicitXMLName(valueField); fieldName == "" {
+				fieldName = typeField.Name
+			}
+		}
+
+		// once the next elem in the path is found, restart with it as root
+		if fieldName == path[0] {
+			if len(path) == 1 {
+				return valueField, nil
+			}
+
+			return getFieldFromPath(valueField, path[1:])
 		}
 	}
+
 	return reflect.Value{}, errFieldNotFound
 }
 
-func getFieldFromIncludePath(val reflect.Value, path []string) (reflect.Value, error) {
-	if path[0] != "Body" {
-		return reflect.Value{}, fmt.Errorf("invalid Include path (should start with Body): %v", path)
+// Unwrap value as much as possible. A value can no longer be unwrapped if:
+// - it is an empty array or slice
+// - it is a nil pointer
+// This assumes, if it encounters an array field, that it will be filling in the first field and continuing
+// TODO: support arbitrary array size.
+func unwrapValue(val reflect.Value) reflect.Value {
+	// if the value is an interface or pointer, get its value
+	if val.Type().Kind() == reflect.Ptr || val.Type().Kind() == reflect.Interface {
+		// if the value is a nil pointer
+		if val.IsNil() {
+			return val
+		}
+
+		return unwrapValue(val.Elem())
 	}
 
-	// WARNING: Dragons below.
-	// We iterate the Content struct.
-	// We attempt to confirm that this is the proper object to store the data in, by
-	// examining the XMLName field and confirming that it matches the first element found in the XOP include path.
-	// Next, we attempt to find the field identified by the subsequent elements found in the XOP include path
-	// by looking for the XML tag that matches then repeating until we reach the end of the include path.
-	// We use HasPrefix, instead of ==, because often the tag includes ',omitempty' as a trailer.
-	// Once we've got it, we return the element.
-
-	rXMLName, err := getFieldByName(val.Elem(), "XMLName")
-	if err != nil {
-		return rXMLName, err
+	// if the value is an array or a slice, assume that we are looking for its first element
+	if val.Type().Kind() == reflect.Slice || val.Type().Kind() == reflect.Array {
+		// if the value is an empty array or slice
+		if val.Cap() == 0 {
+			return val
+		}
+
+		return unwrapValue(val.Index(0))
 	}
 
-	// The XMLName field has 2 properties; one called 'Space' and one called 'Local'.
-	// We only care about 'Local' right now.
-	// We assume the namespace is correct if the local name is correct.
-	for i := 0; i < rXMLName.NumField(); i++ {
-		rXMLTypeField := rXMLName.Type().Field(i)
-		rXMLValueField := rXMLName.Field(i)
+	// the value has been unwrapped
+	return val
+}
 
-		if rXMLTypeField.Name == "Local" && rXMLValueField.String() != path[1] {
-			return reflect.Value{}, fmt.Errorf("invalid XML object path, expected %s, got %s", rXMLValueField.String(), path[1])
-		}
+func getNameFromTag(tag string) string {
```

Comment:
It may be useful to add a comment here with an example field you're trying to handle.

Reply:
I added a comment and a test case